### DISCUSSION TITLE
Support for byte array

### DISCUSSION THREAD

**Discussion by jlack1987** [foxglove/community#1042] [Q&A]
I see that a [packed element field](https://github.com/foxglove/schemas/blob/main/schemas/proto/foxglove/PackedElementField.proto) is supported, but I don't see a messages definition for a byte array. My use case is I just have N time series signals all with their own string name and a p.o.d type to represent their value and i'd like to use foxglove to visualize the signals. I know I could use json or something else, but I'm mainly interested in the byte array for this bc it's a more efficient way of sending the data. Is this already supported or could it be added?

**Comment by defunctzombie** [comment]
That packed element field is specific to PointCloud use and not for general message packing. Based on what you describe it seems like you should be looking at something like protobuf or flatbuffer to store your data in a binary serialized form. You won't need to store the string names and it will behave more like a POD type.

**Reply by jlack1987** [reply]
I appreciate your reply, but I'm not sure I fully understand. If I don't send the names then how would foxglove know which signal is what when I want to plot them?

**Reply by defunctzombie** [reply]
The schema is sent before any messages (whether you are using mcap or websocket protocol) and Foxglove reads and understands the schema.

**Reply by jlack1987** [reply]
That seems like what I want, i'm just not sure how to make that happen based on the docs i've found online for using, for example, the websocket. If for example I had a data message defined as the following
```protobuf
syntax = "proto3";

package pkg;

// A vector in 3D space that represents a direction only
message Data {
  repeated double dbl_signals = 1;
  repeated int32 int_signals = 2;
  repeated uint64 uint_signals = 3;
}
```
Would I be able to use a message like this and visualize the signals in foxglove? I was under the impression that a message like this couldn't be used bc there would be no way for foxglove to know what the name of each element in each vector/list should be named. I'm not super experienced in this area so I may be asking a stupid question :grimacing:

**Reply by defunctzombie** [reply]
Here's an example of a python server to send protobuf data: https://github.com/foxglove/ws-protocol/blob/main/python/src/foxglove_websocket/examples/protobuf_server.py and one in c++ https://github.com/foxglove/ws-protocol/blob/main/cpp/examples/src/example_server_protobuf.cpp

The thing to notice is that the server provides the protobuf description to the app: https://github.com/foxglove/ws-protocol/blob/main/python/src/foxglove_websocket/examples/protobuf_server.py#L60-L70

That's how the app knows the name of the fields.

**Reply by jlack1987** [reply]
Yeah I have seen those. I think those display a different use case than I am working on though because I have `N` signals, all represented by a string name which isn't known until config time, and, for example, a double value. The example message I show up there I think is just not possible given what is currently supported. I'm ultimately just looking for an efficient way of sending a bunch of signals at once and i'm not seeing how to do it. I currently have it functional where it sends it all as a json string, but it's really not an efficient way to send a bunch of time series data. A similar-ish ros2 message definition of what i'm talking about is given [here](https://github.com/PickNikRobotics/data_tamer/blob/main/data_tamer_msgs/msg/Snapshot.msg), but I don't want to go through ROS i'd rather go straight through the foxglove websocket.

**Comment by jlack1987** [comment]
I think it's ok to close this. I have developed a [ros2 package](https://gitlab.com/jlack/data_tamer_tools) that works alongside [data tamer](https://github.com/PickNikRobotics/data_tamer) to provide data logging utilities that both works with ros2 and are as data efficient as I could make them(any ideas for further improving efficiency welcome :grin: ).
